kernel/src/engine/arrow_expression/apply_schema.rs

```diff
@@ -0,0 +1,170 @@
+use super::super::arrow_utils::make_arrow_error;
+use crate::arrow::array::{
+    Array, ArrayRef, AsArray, ListArray, MapArray, RecordBatch, StructArray,
+};
+use crate::arrow::datatypes::{DataType as ArrowDataType, Field as ArrowField};
+use crate::engine::ensure_data_types::ensure_data_types;
+use crate::error::{DeltaResult, Error};
+use crate::schema::{ArrayType, DataType, MapType, Schema, StructField};
+use itertools::Itertools;
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+// Apply a schema to an array. The array _must_ be a `StructArray`. Returns a `RecordBatch where the
+// names of fields, nullable, and metadata in the struct have been transformed to match those in
+// schema specified by `schema`
+pub fn apply_schema(array: &dyn Array, schema: &DataType) -> DeltaResult<RecordBatch> {
```

Comment:
```suggestion
pub(crate) fn apply_schema(array: &dyn Array, schema: &DataType) -> DeltaResult<RecordBatch> {
```
(otherwise it becomes part of the crate's public API where previously it was a private internal function)